foxai/explainer/fullgradeints.py

```diff
@@ -0,0 +1,462 @@
+
+        Returns:
+            bias if present in module else None
+        """
+
+        if isinstance(module, torch.nn.BatchNorm2d):
+            if module.running_var is None or module.running_mean is None:
+                return None
+            b = (
+                -(
+                    module.running_mean
+                    * module.weight
+                    / torch.sqrt(module.running_var + module.eps)
+                )
+                + module.bias
+            )
+            return b.data
+        elif module.bias is None:
+            return None
+        else:
+            return module.bias.data
+
+    @property
+    def biases(self) -> List[torch.Tensor]:
+        """Get biases."""
+        return self._biases
+
+    def _extract_layer_grads(  # pylint: disable = unused-argument
+        self, module, in_grad: _grad_t, out_grad: _grad_t
+    ) -> None:
+        """Collect gradient outputs from each layer, that contains bias term.
+
+        Args:
+            in_grad: input gradients
+            out_grad: output gradients
+        """
+
+        if not module.bias is None:
+            self.feature_grads.append(out_grad[0])
+
+    def get_feature_grads(
+        self, x: torch.Tensor, output_scalar: torch.Tensor
+    ) -> Tuple[torch.Tensor, List[torch.Tensor]]:
+        """Get gradients for the input and for each model layer w.r.t. input.
+
+        Args:
+            x: the input tensor
+            output_scalar: the output tensor
+
+        Returns:
+            gradients for input
+            gradients for each layer, that has bias term
+        """
+
+        # Empty feature grads list
+        self.feature_grads = []
+
+        self.model.zero_grad()
+        # Gradients w.r.t. input
+        input_gradients = torch.autograd.grad(outputs=output_scalar, inputs=x)[0]
+
+        return input_gradients, self.feature_grads
+
+
+class FullGrad:
+    """
+    Compute FullGrad saliency map and full gradient decomposition
+    """
+
+    def __init__(self, model, image_size=(3, 224, 224)):
+        self.model: torch.nn.Module = model
+        self.im_size: Tuple[int, int, int, int] = (1,) + image_size
+        self.model_ext: FullGradExtractor = FullGradExtractor(
+            model=self.model, im_size=image_size
+        )
+        self._biases: List[torch.Tensor] = self.model_ext.biases
+
+        cuda = next(self.model.parameters()).is_cuda
+        self._device = torch.device("cuda" if cuda else "cpu")
+        _ = self.check_completeness()
+
+    def check_completeness(self) -> bool:
+        """Check if completeness property is satisfied. If not, it usually means that
+        some bias gradients are not computed (e.g.: implicit biases of non-linearities).
+
+        Returns:
+            whether the completnes condition is fulfielled
+        """
+
+        # Random input image
+        sample_input: torch.Tensor = torch.randn(self.im_size).to(self._device)
+
+        # Get raw outputs
+        self.model.eval()
```

Comment:
In this line, we are modifying the model's state. We should restore somehow restore it at the end.

Reply:
done